core/src/main/kotlin/org/evomaster/core/problem/api/service/ApiWsStructureMutator.kt

```diff
@@ -25,56 +27,132 @@ import kotlin.math.min
+        val esr = individual.fitness.getViewAccessedExternalServiceRequests()
+        if (esr.isEmpty()) {
+            //nothing to do
+            return
+        }
 
-            val actions = mutableListOf<ExternalServiceAction>().plus(
-                sampler.getExternalService().getExternalServiceActions()
-            )
+        val newActions: MutableList<HttpExternalServiceAction> = mutableListOf()
 
-            ind.addInitializingDbActions(0, actions)
+        ind.seeMainExecutableActions().forEachIndexed { index, action ->
+            val parent = action.parent
+            if (parent !is EnterpriseActionGroup) {
+                //TODO this should not really happen
+                val msg = "Action is not inside an EnterpriseActionGroup"
+                log.error(msg)
+                throw RuntimeException(msg)
+            }
 
-            if (log.isTraceEnabled)
-                log.trace("{} existingExternalServiceData are added", actions)
+            // Adding the new [HttpExternalServiceAction] will be handled here. Handling
+            // used and not used external service actions will be handled in Fitness using
+            // [used] property in action.
+            if (esr.containsKey(index)) {
+                val requests = esr[index]
+
+                if (requests!!.isNotEmpty()) {
+                    val existingActions = parent.getExternalServiceActions()
+
+                    val actions: MutableList<HttpExternalServiceAction> = mutableListOf()
+
+                    requests
+                        .groupBy { it.absoluteURL }
+                        .forEach { (url, grequests) ->
+                            // here, we assume that the front external service actions should be accessed
+                            val startingIndex = existingActions.filterIsInstance<HttpExternalServiceAction>().count { it.request.absoluteURL == url}
+                            if (startingIndex < grequests.size){
+                                (startingIndex until  grequests.size).forEach {i->
+                                    val a = externalServiceHandler
+                                        .createExternalServiceAction(grequests[i])
+                                    a.confirmUsed()
+                                    actions.add(a)
+                                }
+                            }
+                    }
+
+                    if (actions.isNotEmpty()) {
+                        newActions.addAll(actions)
+                        parent.addChildrenToGroup(
+                            actions,
+                            GroupsOfChildren.EXTERNAL_SERVICES
+                        )
+                    }
+                }
+            }
+        }
+
+        // all actions should have local ids
+        Lazy.assert {
+            ind.seeAllActions().all { it.hasLocalId() }
+        }
+
+        if (log.isTraceEnabled)
+            log.trace("{} existingExternalServiceData are added")
```

Comment:
why the `{}` here? you are not interpolating anything in the log. did u miss a `, x.size())`  as second parameter of the `trace`?

Reply:
Yes, size of the newly added actions. Fixed now.